chatlas/_chat.py

```diff
@@ -298,15 +292,91 @@ def tokens(
+                        "tokens_total": tj.tokens[1],
+                    },
                 ]
             )
 
         return res
 
+    def get_cost(
+        self,
+        options: CostOptions = "all",
+        token_price: Optional[tuple[float, float]] = None,
+    ) -> float:
+        """
+        Get the cost of the chat. Note that this is a rough estimate. Providers may change their pricing frequently and without notice.
+
+        Parameters
+        ----------
+        options
+            One of the following (default is "all"):
+              - `"all"`: Return the total cost of all turns in the chat.
+              - `"last"`: Return the cost of the last turn in the chat.
+        token_price
+            An optional tuple in the format of (input_token_cost, output_token_cost) for bringing your own cost information.
+                 - `"input_token_cost"`: The cost per user token in USD per million tokens.
+                 - `"output_token_cost"`: The cost per assistant token in USD per million tokens.
+
+        Returns
+        -------
+        float
+            The cost of the chat, in USD.
+        """
+
+        # Look up token cost for user and input tokens based on the provider and model
+        turns_tokens = self.get_tokens()
+        if token_price:
+            input_token_price = token_price[0] / 1e6
+            output_token_price = token_price[1] / 1e6
+        else:
+            price_token = get_token_pricing(self.provider.name, self.provider.model)
+            input_token_price = price_token["input"] / 1e6
+            output_token_price = price_token["output"] / 1e6
+
+        if not input_token_price and not output_token_price:
+            raise KeyError(
+                f"We could not locate provider ' { self.provider.name } ' and model '{ self.provider.model } ' in our pricing information. Please supply your own if you wish to use the cost function."
+            )
+
+        if len(turns_tokens) == 0:
+            return 0.0
+
+        if options == "last":
+            last_turn = turns_tokens[len(turns_tokens) - 1]
+            acc = 0.0
+            if last_turn["role"] == "assistant":
+                acc += last_turn["tokens"] * output_token_price
+            elif last_turn["role"] == "user":
+                acc += last_turn["tokens_total"] * input_token_price
+            else:
+                raise ValueError(f"Unrecognized role type { last_turn['role'] }")
+            return acc
+
+        if options == "all":
+            asst_tokens = sum(
+                u["tokens_total"] for u in turns_tokens if u["role"] == "assistant"
+            )
+            user_tokens = sum(
+                u["tokens_total"] for u in turns_tokens if u["role"] == "user"
+            )
+            cost = (asst_tokens * output_token_price) + (
+                user_tokens * input_token_price
+            )
+            return cost
+
```

Comment:
Could raise an error about unsupported `options` value here.

In general, I'm usually not a fan of throwing run time errors about values that don't pass type checks, but in this case, maybe it's worth it?